### DISCUSSION TITLE
Problems extending PhantomCamera3D

### DISCUSSION THREAD

**Discussion by stevenmraines** [ramokz/phantom-camera#363] [Question / Help Wanted] [answered]
I'm trying to extend PhantomCamera3D with some extra functionality, mainly the sort of swaying POV kind of motion seen in [this video](https://youtu.be/rwdikOjFIQA?list=PLf1ulySF2JttZBEKe1WX9_Eeap4IuE4Yv&t=34). I made this script that extends PhantomCamera3D:

```
extends PhantomCamera3D

@export var camera_sway_noise_amplitude = 1.0
@export var camera_sway_noise_frequency = 10.0
@export var sway := true

var camera_sway_x_noise
var camera_sway_y_noise
var camera_sway_noise_rotation = Vector3()
var camera_sway_noise_count = 0


func _ready():
	camera_sway_x_noise = FastNoiseLite.new()
	camera_sway_x_noise.noise_type = FastNoiseLite.TYPE_SIMPLEX
	camera_sway_x_noise.seed = 0
	camera_sway_x_noise.fractal_octaves = 1
	
	camera_sway_y_noise = FastNoiseLite.new()
	camera_sway_y_noise.noise_type = FastNoiseLite.TYPE_SIMPLEX
	camera_sway_y_noise.seed = 1
	camera_sway_y_noise.fractal_octaves = 1


func _physics_process(delta):
	if ! sway:
		return

	camera_sway_noise_count += delta * camera_sway_noise_frequency
	camera_sway_noise_rotation.x = camera_sway_x_noise.get_noise_1d(camera_sway_noise_count) \
		* camera_sway_noise_amplitude
	camera_sway_noise_rotation.y = camera_sway_y_noise.get_noise_1d(camera_sway_noise_count) \
		* camera_sway_noise_amplitude
	if look_at_mode == PhantomCamera3D.LookAtMode.MIMIC:
		look_at_target.rotation_degrees = camera_sway_noise_rotation
	elif look_at_mode == PhantomCamera3D.LookAtMode.SIMPLE and look_at_target:
		look_at_target.position = camera_sway_noise_rotation
```

And I made a new scene called sway_camera that just has a Node3D with the above script attached. This was working well enough for cutscenes (mouse movement to look around when not in cutscenes was causing issues, but that's alright because I only need it for cutscenes). However, I'm trying to animate my custom camera and the normal PhantomCamera3D functionality isn't working as expected. It has a follow mode of glued with a target of some Node3D, and it's look at mode is simple with an initial value of some different Node3D. One second after my extended camera is made active in the cutscene, my animation changes it's look at target to a path follower.

Strangely, the initial Node3D look at target is not being looked at. But when the animation changes the target, it correctly points at the path follower. Maybe it's because the initial target is being set in the inspector and the second one is being set by the animation which calls a method to change it with set_look_at_target? I also noticed that the follow offset didn't seem to be working at all.

Is there maybe some manual setup I'll have to do in my extended camera script? Maybe PhantomCameraManager isn't picking up on my extended camera when it bootstraps everything (or whatever it does, I haven't looked at that code yet)?

EDIT: After playing around with it some more, it looks like it WAS pointing at something with the initial look at target, it was just not pointing in the expected direction. It's like it's off by about 45 degrees on the Y axis.

**Comment by ramokz** [comment, marked as answer]
It's worth noting that what it looks like you're trying to add is what is planned in the next release (#126).
So this might become redundant once v0.8 is out.

Beyond that, there are a few things worth mentioning about the setup you shared.

Firstly, is the `PCam3D` logic occurring in the `_physics_process` or `_process`? The addon will dynamically use one or the other depending on its target(s), and so you too would need to dynamically use one or the other depending on which is being used in the `PCam3D` script. At the minute, and ideally, it should be the `_process` for `PCam3D` in 3D scenes.

Secondly, and unrelated to the first, are you looking to retain the logic that is happening in the `_physics_process` from `phantom_camera_3d.gd` script? If so, you need to call `super(delta)` in order for that to work. Without it, your extended script overrides and ignores the original `PCam3D` function, and it won't do anything with the `look_at_position` property.

Thirdly, modifying the `look_at_target` value in your script ought to not work in the way you've set it up, since that property is both set and [used](https://github.com/ramokz/phantom-camera/blob/3b9101eeb5c2fdfd380198629e5b6bf6b45a409f/addons/phantom_camera/scripts/phantom_camera/phantom_camera_3d.gd#L789) to change the rotation within the `PCam3D` script itself. So you're effectively modifying a property that has already been used to set the camera rotation by the time you're doing anything in your extended script.

Honestly, I don't see a simple way of doing what you're looking to do without modifying the addon's source files. Not impossible by any means, but reckon it's not worth the effort given, again, camera noise is planned for the next release.

**Reply by stevenmraines** [reply]
Thanks, guess I'll just wait for v0.8 then.
